File: R/05_round_trips.R
# 05_round_trips.R
# Cyclistic Bike-Share Case Study — R Module
#
# Round trip rate by rider type.
# Confirms: casual riders return to starting station 1.63x more often than members.
#
# Introduces: is.na(), coalesce(), case_when()

library(tidyverse)

csv_folder <- file.path(dirname(dirname(rstudioapi::getSourceEditorContext()$path)),
                        "Data", "DATAfile_Consolidated")
csv_files  <- list.files(csv_folder, pattern = "\\.csv$", full.names = TRUE)
trips      <- read_csv(csv_files, show_col_types = FALSE)

# ---------------------------------------------------------------------------
# Classify round trips
#
# A round trip is where the start and end station are the same.
# Key NULL semantics issue — same as we solved in DuckDB and SQL Server:
#
#   NA == NA  evaluates to NA in R (not TRUE) — same as SQL's NULL = NULL
#
# To replicate Power BI's BLANK() = BLANK() = TRUE behavior, we replace
# NA with an empty string before comparing, using coalesce().
# coalesce() returns the first non-NA value — R's equivalent of SQL COALESCE()
# and DuckDB's COALESCE().
#
# This ensures e-bike rides with no station name (locked at non-station racks)
# are correctly counted as round trips when both ends are missing.
# ---------------------------------------------------------------------------

trips <- trips |>
  mutate(
    is_round_trip = coalesce(start_station_name, "") == coalesce(end_station_name, "")
  )

# ---------------------------------------------------------------------------
# Summarise round trip rates
# ---------------------------------------------------------------------------

round_trip_summary <- trips |>
  group_by(member_casual) |>
  summarise(
    total_rides     = n(),
    round_trips     = sum(is_round_trip),
    round_trip_pct  = round(round_trips / total_rides * 100, 1)
  )

cat("--- Round Trip Rates ---\n")
print(round_trip_summary)

casual_pct <- round_trip_summary$round_trip_pct[round_trip_summary$member_casual == "casual"]
member_pct <- round_trip_summary$round_trip_pct[round_trip_summary$member_casual == "member"]

cat("\nRound trip ratio (casual / member):", round(casual_pct / member_pct, 2), "\n")

# ---------------------------------------------------------------------------
# Bar chart
# ---------------------------------------------------------------------------

ggplot(round_trip_summary, aes(x = member_casual, y = round_trip_pct, fill = member_casual)) +
  geom_col(width = 0.5) +
  geom_text(
    aes(label = paste0(round_trip_pct, "%")),
    vjust = -0.4,
    size  = 4
  ) +
  scale_fill_manual(values = c("casual" = "#F4A261", "member" = "#2A9D8F")) +
  scale_y_continuous(limits = c(0, 25), labels = scales::percent_format(scale = 1)) +
  labs(
    title    = "Casual Riders Return to Their Starting Station Far More Often",
    subtitle = "Casual round trip rate is 1.63x higher — consistent with recreational riding",
    x        = NULL,
    y        = "Round Trip Rate",
    fill     = NULL,
    caption  = "Source: Divvy Trip Data"
  ) +
  theme_minimal(base_size = 13) +
  theme(legend.position = "none")

ggsave(
  file.path(dirname(dirname(rstudioapi::getSourceEditorContext()$path)),
            "Visuals", "R", "05_round_trips.png"),
  width = 8, height = 6, dpi = 150
)

File: Python/pandas/05_round_trips.py
"""
05_round_trips.py
-----------------
Do casual riders return to their starting station more often than members,
indicating recreational use?

Key finding: Casual riders take round trips at 1.63x the rate of members —
further evidence of leisure riding patterns.

pandas notes:
  - NaN semantics in pandas mirror SQL's three-valued logic: NaN == NaN
    evaluates to False, just like NULL = NULL → UNKNOWN in SQL. A naive
    comparison of start_station_name == end_station_name would silently
    exclude all e-bike rides where both station names are missing.
  - Fix: .fillna('') converts NaN to empty string first, so two missing
    station names compare as equal — matching Power BI DAX's behavior where
    BLANK() = BLANK() is TRUE.
  - This is the same fix applied in DuckDB (COALESCE) and SQL Server
    (ISNULL) — the pandas idiom just looks different.

Expected output (matching SQL Server, DuckDB, and R):
    Member round trip %:  ~11.1%
    Casual round trip %:  ~18.1%
    Round trip ratio (casual / member): 1.63x
"""

import pandas as pd
from utils import load_trips

trips = load_trips()

# ---------------------------------------------------------------------------
# Round trip flag
#
# fillna('') converts NaN → '' before comparison so that rides where both
# station names are missing (e-bikes at non-station racks) correctly count
# as round trips — matching the Power BI DAX BLANK() = BLANK() behavior.
# ---------------------------------------------------------------------------

trips["is_round_trip"] = (
    trips["start_station_name"].fillna("") == trips["end_station_name"].fillna("")
)

# ---------------------------------------------------------------------------
# Per-group round trip counts and percentages
# ---------------------------------------------------------------------------

group = trips.groupby("member_casual")["is_round_trip"].agg(
    total_rides="count",
    round_trips="sum"
)
group["round_trip_pct"] = group["round_trips"] / group["total_rides"] * 100

total_rt   = trips["is_round_trip"].sum()
member_pct = group.loc["member", "round_trip_pct"]
casual_pct = group.loc["casual", "round_trip_pct"]
member_rt  = group.loc["member", "round_trips"]
casual_rt  = group.loc["casual", "round_trips"]
ratio      = casual_pct / member_pct

print("=" * 55)
print("  Round Trip Analysis")
print("=" * 55)
print(f"  Total round trips:  {total_rt:,}")
print()
print(f"  {'':30s} {'Round Trip %':>12}")
print(f"  {'-'*48}")
print(f"  {'Members':30s} {member_pct:>11.1f}%  ({member_rt:,})")
print(f"  {'Casual riders':30s} {casual_pct:>11.1f}%  ({casual_rt:,})")
print(f"  {'-'*48}")
print(f"  Round trip ratio (casual/member): {ratio:.2f}x")
print("=" * 55)
print("  Note: NaN station names treated as equal (.fillna('')),")
print("  matching Power BI DAX BLANK() = BLANK() behavior.")
print("  e-bike rides ending at non-station racks count as")
print("  round trips when start and end locations both missing.")
print("=" * 55)
